File: Plugins/Kingbotk/AWB Plugin/WikiProjects/WPIndia.vb
Friend NotInheritable Class WPIndia
    Inherits PluginBase

    ' Settings:
    Private OurTab As New TabPage("India")
    Private WithEvents OurSettingsControl As New WPIndiaSettings
    Private Const conEnabled As String = "IndEnabled"

    Protected Friend Overrides ReadOnly Property PluginShortName() As String
        Get
            Return "India"
        End Get
    End Property

    Protected Overrides Sub ImportanceParameter(ByVal Importance As Importance)
        Template.NewOrReplaceTemplateParm("importance", Importance.ToString, Me.Article, False, False)
    End Sub
    Protected Friend Overrides ReadOnly Property GenericSettings() As IGenericSettings
        Get
            Return OurSettingsControl
        End Get
    End Property
    Protected Overrides ReadOnly Property CategoryTalkClassParm() As String
        Get
            Return "Cat"
        End Get
    End Property
    Protected Overrides ReadOnly Property TemplateTalkClassParm() As String
        Get
            Return "Template"
        End Get
    End Property
    Protected Overrides ReadOnly Property PreferredTemplateName() As String
        Get
            Return "WikiProject India"
        End Get
    End Property
    Friend Overrides ReadOnly Property HasReqPhotoParam() As Boolean
        Get
            Return True
        End Get
    End Property
    Friend Overrides Sub ReqPhoto()
        AddNewParamWithAYesValue("image-needed")
    End Sub
    'Protected Overrides ReadOnly Property PreferredTemplateNameRegexString() As String
    '    Get
    '        Return "^[Ww]P India$"
    '    End Get
    'End Property

    ' Initialisation:
    Friend Sub New()
        MyBase.New("") ' Specify alternate names only
    End Sub
    Protected Friend Overrides Sub Initialise()
        OurMenuItem = New ToolStripMenuItem("India Plugin")
        MyBase.InitialiseBase() ' must set menu item object first
        OurTab.UseVisualStyleBackColor = True
        OurTab.Controls.Add(OurSettingsControl)
    End Sub

    ' Article processing:
    Protected Overrides ReadOnly Property InspectUnsetParameters() As Boolean
        Get
            Return False
        End Get
    End Property
    Protected Overrides Sub InspectUnsetParameter(ByVal Param As String)
    End Sub
    Protected Overrides Function SkipIfContains() As Boolean
        ' None
    End Function
    Protected Overrides Sub ProcessArticleFinish()
        StubClass()
        With OurSettingsControl
            If .Geography Then
                AddAndLogNewParamWithAYesValue("geography")
                AddEmptyParam("geography-importance")
            End If
            If .States Then
                AddAndLogNewParamWithAYesValue("states")
                AddEmptyParam("states-importance")
            End If
            If .Districts Then
                AddAndLogNewParamWithAYesValue("districts")
                AddEmptyParam("districts-importance")
            End If
            If .Cities Then
                AddAndLogNewParamWithAYesValue("cities")
                AddEmptyParam("cities-importance")
            End If
            If .Maps Then
                AddAndLogNewParamWithAYesValue("maps")
            End If
            If .Cinema Then
                AddAndLogNewParamWithAYesValue("cinema")
                AddEmptyParam("cinema-importance")
            End If
            If .History Then
                AddAndLogNewParamWithAYesValue("history")
                AddEmptyParam("history-importance")
            End If
            If .Literature Then
                AddAndLogNewParamWithAYesValue("literature")
                AddEmptyParam("literature-importance")
            End If
            If .Politics Then
                AddAndLogNewParamWithAYesValue("politics")
                AddEmptyParam("politics-importance")
            End If
            If .ProtectedAreas Then
                AddAndLogNewParamWithAYesValue("protected-areas")
                AddEmptyParam("protected-areas-importance")
            End If
            If .Tamil Then
                AddAndLogNewParamWithAYesValue("tamil")
                AddEmptyParam("tamil-importance")
            End If
            If .Television Then
                AddAndLogNewParamWithAYesValue("television")
                AddEmptyParam("television-importance")
            End If
            If .Andhra Then
                AddAndLogNewParamWithAYesValue("andhra")
                AddEmptyParam("andhra-importance")
            End If
            If .Arunachal Then
                AddAndLogNewParamWithAYesValue("arunachal")
                AddEmptyParam("arunachal-importance")
            End If
            If .Assam Then
                AddAndLogNewParamWithAYesValue("assam")
                AddEmptyParam("assam-importance")
            End If
            If .Bengal Then
                AddAndLogNewParamWithAYesValue("bengal")
                AddEmptyParam("bengal-importance")
            End If
            If .Bihar Then
                AddAndLogNewParamWithAYesValue("bihar")
                AddEmptyParam("bihar-importance")
            End If
            If .Chhattisgarh Then
                AddAndLogNewParamWithAYesValue("chhattisgarh")
                AddEmptyParam("chhattisgarh-importance")
            End If
            If .Goa Then
                AddAndLogNewParamWithAYesValue("goa")
                AddEmptyParam("goa-importance")
            End If
            If .Gujarat Then
                AddAndLogNewParamWithAYesValue("gujarat")
                AddEmptyParam("gujarat-importance")
            End If
            If .Haryana Then
                AddAndLogNewParamWithAYesValue("haryana")
                AddEmptyParam("haryana-importance")
            End If
            If .Himachal Then
                AddAndLogNewParamWithAYesValue("himachal")
                AddEmptyParam("himachal-importance")
            End If
            If .JandK Then
                AddAndLogNewParamWithAYesValue("jandk")
                AddEmptyParam("jandk-importance")
            End If
            If .Jharkhand Then
                AddAndLogNewParamWithAYesValue("jharkhand")
                AddEmptyParam("jharkhand-importance")
            End If
            If .Kerala Then
                AddAndLogNewParamWithAYesValue("kerala")
                AddEmptyParam("kerala-importance")
            End If
            If .Karnataka Then
                AddAndLogNewParamWithAYesValue("karnataka")
                AddEmptyParam("karnataka-importance")
            End If
            If .Madhya Then
                AddAndLogNewParamWithAYesValue("madhya")
                AddEmptyParam("madhya-importance")
            End If
            If .Maharashtra Then
                AddAndLogNewParamWithAYesValue("maharashtra")
                AddEmptyParam("maharashtra-importance")
            End If
            If .Manipur Then
                AddAndLogNewParamWithAYesValue("manipur")
                AddEmptyParam("manipur-importance")
            End If
            If .Meghalaya Then
                AddAndLogNewParamWithAYesValue("meghalaya")
                AddEmptyParam("meghalaya-importance")
            End If
            If .Mizoram Then
                AddAndLogNewParamWithAYesValue("mizoram")
                AddEmptyParam("mizoram-importance")
            End If
            If .Nagaland Then
                AddAndLogNewParamWithAYesValue("nagaland")
                AddEmptyParam("nagaland-importance")
            End If
            If .Orissa Then
                AddAndLogNewParamWithAYesValue("orissa")
                AddEmptyParam("orissa-importance")
            End If
            If .Punjab Then
                AddAndLogNewParamWithAYesValue("punjab")
                AddEmptyParam("punjab-importance")
            End If
            If .Rajasthan Then
                AddAndLogNewParamWithAYesValue("rajasthan")
                AddEmptyParam("rajasthan-importance")
            End If
            If .Sikkim Then
                AddAndLogNewParamWithAYesValue("sikkim")
                AddEmptyParam("sikkim-importance")
            End If
            If .Tamilnadu Then
                AddAndLogNewParamWithAYesValue("tamilnadu")
                AddEmptyParam("tamilnadu-importance")
            End If
            If .Tripura Then
                AddAndLogNewParamWithAYesValue("tripura")
                AddEmptyParam("tripura-importance")
            End If
            If .Uttar Then
                AddAndLogNewParamWithAYesValue("uttar")
                AddEmptyParam("uttar-importance")
            End If
            If .Uttarakand Then
                AddAndLogNewParamWithAYesValue("uttarakand")
                AddEmptyParam("uttarakand-importance")
            End If
            If .Chennai Then
                AddAndLogNewParamWithAYesValue("chennai")
                AddAndLogEmptyParam("chennai-importance")
            End If
            If .Mumbai Then
                AddAndLogNewParamWithAYesValue("mumbai")
                AddAndLogEmptyParam("mumbai-importance")
            End If
        End With
    End Sub
    Protected Overrides Function TemplateFound() As Boolean
    End Function
    Protected Overrides Sub GotTemplateNotPreferredName(ByVal TemplateName As String)
    End Sub
    Protected Overrides Function WriteTemplateHeader(ByRef PutTemplateAtTop As Boolean) As String
        WriteTemplateHeader = "{{WikiProject India India" & _
           Microsoft.VisualBasic.vbCrLf & WriteOutParameterToHeader("class") & _
           WriteOutParameterToHeader("importance")
    End Function

    'User interface:
    Protected Overrides Sub ShowHideOurObjects(ByVal Visible As Boolean)
        PluginManager.ShowHidePluginTab(OurTab, Visible)
    End Sub

    ' XML settings:
    Protected Friend Overrides Sub ReadXML(ByVal Reader As System.Xml.XmlTextReader)
        Dim blnNewVal As Boolean = PluginManager.XMLReadBoolean(Reader, conEnabled, Enabled)
        If Not blnNewVal = Enabled Then Enabled = blnNewVal ' Mustn't set if the same or we get extra tabs
        OurSettingsControl.ReadXML(Reader)
    End Sub
    Protected Friend Overrides Sub Reset()
        OurSettingsControl.Reset()
    End Sub
    Protected Friend Overrides Sub WriteXML(ByVal Writer As System.Xml.XmlTextWriter)
        Writer.WriteAttributeString(conEnabled, Enabled.ToString)
        OurSettingsControl.WriteXML(Writer)
    End Sub
End Class

File: Plugins/Kingbotk Plugin/WikiProjects/WPIndia.cs
/*
Copyright (C) 2008 Stephen Kennedy (Kingboyk) http://www.sdk-software.com/
Copyright (C) 2008 Sam Reed (Reedy) http://www.reedyboy.net/

This program is free software; you can redistribute it and/or modify it under the terms of Version 2 of the GNU General Public License as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License Version 2 along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

using System.Windows.Forms;
using System.Xml;

namespace AutoWikiBrowser.Plugins.Kingbotk.WikiProjects
{
    internal sealed class WPIndia : PluginBase
    {
        // Settings:
        private readonly TabPage _ourTab = new TabPage("India");

        private readonly GenericWithWorkgroups _ourSettingsControl;
        private const string Prefix = "Ind";

        private const string PluginName = "WikiProject India";
        private const string GeographyGroup = "Geography";

        private const string OthersGroup = "Others";

        private readonly TemplateParameters[] _params =
        {
            new TemplateParameters
            {
                StorageKey = "Geography",
                Group = GeographyGroup,
                ParamName = "Geography"
            },
            new TemplateParameters
            {
                StorageKey = "Maps",
                Group = GeographyGroup,
                ParamName = "Maps"
            },
            new TemplateParameters
            {
                StorageKey = "Cities",
                Group = GeographyGroup,
                ParamName = "Cities"
            },
            new TemplateParameters
            {
                StorageKey = "Chennai",
                Group = GeographyGroup,
                ParamName = "Chennai"
            },
            new TemplateParameters
            {
                StorageKey = "Mumbai",
                Group = GeographyGroup,
                ParamName = "Mumbai"
            },
            new TemplateParameters
            {
                StorageKey = "Hyderabad",
                Group = GeographyGroup,
                ParamName = "Hyderabad"
            },
            new TemplateParameters
            {
                StorageKey = "Districts",
                Group = GeographyGroup,
                ParamName = "Districts"
            },
            new TemplateParameters
            {
                StorageKey = "States",
                Group = GeographyGroup,
                ParamName = "States"
            },
            new TemplateParameters
            {
                StorageKey = "Andhra",
                Group = GeographyGroup,
                ParamName = "Andhra Pradesh"
            },
            new TemplateParameters
            {
                StorageKey = "Arunachal",
                Group = GeographyGroup,
                ParamName = "Arunachal Pradesh"
            },
            new TemplateParameters
            {
                StorageKey = "Assam",
                Group = GeographyGroup,
                ParamName = "Assam"
            },
            new TemplateParameters
            {
                StorageKey = "Bengal",
                Group = GeographyGroup,
                ParamName = "Bengal"
            },
            new TemplateParameters
            {
                StorageKey = "Bihar",
                Group = GeographyGroup,
                ParamName = "Bihar"
            },
            new TemplateParameters
            {
                StorageKey = "Chhattisgarh",
                Group = GeographyGroup,
                ParamName = "Chhattisgarh"
            },
            new TemplateParameters
            {
                StorageKey = "Goa",
                Group = GeographyGroup,
                ParamName = "Goa"
            },
            new TemplateParameters
            {
                StorageKey = "Gujarat",
                Group = GeographyGroup,
                ParamName = "Gujarat"
            },
            new TemplateParameters
            {
                StorageKey = "Haryana",
                Group = GeographyGroup,
                ParamName = "Haryana"
            },
            new TemplateParameters
            {
                StorageKey = "Himachal",
                Group = GeographyGroup,
                ParamName = "Himachal Pradesh"
            },
            new TemplateParameters
            {
                StorageKey = "JandK",
                Group = GeographyGroup,
                ParamName = "Jammu and Kashmir"
            },
            new TemplateParameters
            {
                StorageKey = "Jharkhand",
                Group = GeographyGroup,
                ParamName = "Jharkhand"
            },
            new TemplateParameters
            {
                StorageKey = "Karnataka",
                Group = GeographyGroup,
                ParamName = "Karnataka"
            },
            new TemplateParameters
            {
                StorageKey = "Kerala",
                Group = GeographyGroup,
                ParamName = "Kerala"
            },
            new TemplateParameters
            {
                StorageKey = "Madhya",
                Group = GeographyGroup,
                ParamName = "Madhya Pradesh"
            },
            new TemplateParameters
            {
                StorageKey = "Maharashtra",
                Group = GeographyGroup,
                ParamName = "Maharashtra"
            },
            new TemplateParameters
            {
                StorageKey = "Manipur",
                Group = GeographyGroup,
                ParamName = "Manipur"
            },
            new TemplateParameters
            {
                StorageKey = "Meghalaya",
                Group = GeographyGroup,
                ParamName = "Meghalaya"
            },
            new TemplateParameters
            {
                StorageKey = "Mizoram",
                Group = GeographyGroup,
                ParamName = "Mizoram"
            },
            new TemplateParameters
            {
                StorageKey = "Nagaland",
                Group = GeographyGroup,
                ParamName = "Nagaland"
            },
            new TemplateParameters
            {
                StorageKey = "Orissa",
                Group = GeographyGroup,
                ParamName = "Orissa"
            },
            new TemplateParameters
            {
                StorageKey = "Punjab",
                Group = GeographyGroup,
                ParamName = "Punjab"
            },
            new TemplateParameters
            {
                StorageKey = "Rajasthan",
                Group = GeographyGroup,
                ParamName = "Rajasthan"
            },
            new TemplateParameters
            {
                StorageKey = "Sikkim",
                Group = GeographyGroup,
                ParamName = "Sikkim"
            },
            new TemplateParameters
            {
                StorageKey = "Tamilnadu",
                Group = GeographyGroup,
                ParamName = "Tamil Nadu"
            },
            new TemplateParameters
            {
                StorageKey = "Telangana",
                Group = GeographyGroup,
                ParamName = "Telangana"
            },
            new TemplateParameters
            {
                StorageKey = "Tripura",
                Group = GeographyGroup,
                ParamName = "Tripura"
            },
            new TemplateParameters
            {
                StorageKey = "Uttar",
                Group = GeographyGroup,
                ParamName = "Uttar Pradesh"
            },
            new TemplateParameters
            {
                StorageKey = "Uttarakand",
                Group = GeographyGroup,
                ParamName = "Uttarakand"
            },
            new TemplateParameters
            {
                StorageKey = "Andaman",
                Group = GeographyGroup,
                ParamName = "Andaman and Nicobar Island"
            },
            new TemplateParameters
            {
                StorageKey = "Chandigarh",
                Group = GeographyGroup,
                ParamName = "Chandigarh"
            },
            new TemplateParameters
            {
                StorageKey = "Dadra",
                Group = GeographyGroup,
                ParamName = "Dadra and Nagar Haveli"
            },
            new TemplateParameters
            {
                StorageKey = "Daman",
                Group = GeographyGroup,
                ParamName = "Daman and Diu"
            },
            new TemplateParameters
            {
                StorageKey = "Delhi",
                Group = GeographyGroup,
                ParamName = "Delhi"
            },
            new TemplateParameters
            {
                StorageKey = "Lakshadweep",
                Group = GeographyGroup,
                ParamName = "Lakshadweep"
            },
            new TemplateParameters
            {
                StorageKey = "Puducherry",
                Group = GeographyGroup,
                ParamName = "Puducherry"
            },
            new TemplateParameters
            {
                StorageKey = "Cinema",
                Group = OthersGroup,
                ParamName = "Cinema"
            },
            new TemplateParameters
            {
                StorageKey = "History",
                Group = OthersGroup,
                ParamName = "History"
            },
            new TemplateParameters
            {
                StorageKey = "Literature",
                Group = OthersGroup,
                ParamName = "Literature"
            },
            new TemplateParameters
            {
                StorageKey = "Politics",
                Group = OthersGroup,
                ParamName = "Politics"
            },
            new TemplateParameters
            {
                StorageKey = "ProtectedAreas",
                Group = OthersGroup,
                ParamName = "Protected Areas"
            },
            new TemplateParameters
            {
                StorageKey = "Tamil",
                Group = OthersGroup,
                ParamName = "Tamil"
            },
            new TemplateParameters
            {
                StorageKey = "Tele",
                Group = OthersGroup,
                ParamName = "Television"
            }
        };

        internal WPIndia() : base("WPIndia|WP India|WPINDIA|WP INDIA")
        {
            // Specify alternate names only

            _ourSettingsControl = new GenericWithWorkgroups(PluginName, Prefix, true, _params);
        }

        protected internal override string PluginShortName
        {
            get { return "India"; }
        }

        protected override void ImportanceParameter(Importance importance)
        {
            Template.NewOrReplaceTemplateParm("importance", importance.ToString(), TheArticle, false, false);
        }

        protected internal override IGenericSettings GenericSettings
        {
            get { return _ourSettingsControl; }
        }

        protected override string PreferredTemplateName
        {
            get { return PluginName; }
        }

        internal override bool HasReqPhotoParam
        {
            get { return true; }
        }

        internal override void ReqPhoto()
        {
            AddNewParamWithAYesValue("image-needed");
        }

        // Initialisation:
        protected internal override void Initialise()
        {
            OurMenuItem = new ToolStripMenuItem("India Plugin");
            InitialiseBase();
            // must set menu item object first
            _ourTab.UseVisualStyleBackColor = true;
            _ourTab.Controls.Add(_ourSettingsControl);
        }

        // Article processing:
        protected override bool SkipIfContains()
        {
            // None
            return false;
        }

        protected override void ProcessArticleFinish()
        {
            StubClass();
            foreach (ListViewItem lvi in _ourSettingsControl.ListView1.Items)
            {
                if (lvi.Checked)
                {
                    TemplateParameters tp = (TemplateParameters) lvi.Tag;
                    string param = tp.ParamName.ToLower().Replace(" ", "-");
                    AddAndLogNewParamWithAYesValue(param);
                    AddEmptyParam(param + "-importance");
                }
            }
        }

        protected override bool TemplateFound()
        {
            return false;
        }

        protected override string WriteTemplateHeader()
        {
            return "{{" + PluginName + WriteOutParameterToHeader("class") + WriteOutParameterToHeader("importance");
        }

        // User interface:
        protected override void ShowHideOurObjects(bool visible)
        {
            PluginManager.ShowHidePluginTab(_ourTab, visible);
        }

        // XML settings:
        protected internal override void ReadXML(XmlTextReader reader)
        {
            Enabled = PluginManager.XMLReadBoolean(reader, Prefix + "Enabled", Enabled);
            _ourSettingsControl.ReadXML(reader);
        }

        protected internal override void Reset()
        {
            _ourSettingsControl.Reset();
        }

        protected internal override void WriteXML(XmlTextWriter writer)
        {
            writer.WriteAttributeString(Prefix + "Enabled", Enabled.ToString());
            _ourSettingsControl.WriteXML(writer);
        }
    }
}
